src/signing.rs

```diff
@@ -0,0 +1,310 @@
+                .sign(client_secret_key, password, HashAlgorithm::SHA2_256)?
+                .compress(CompressionAlgorithm::ZIP)?,
+        )
+    }
+}
+
+impl TryFrom<&Message> for OpenPGPMessage {
+    type Error = Error;
+
+    fn try_from(m: &Message) -> Result<Self, Self::Error> {
+        Ok(OpenPGPMessage::Literal(LiteralData::from_str(
+            &m.id.to_string(),
+            &serde_json::to_string(&m)?,
+        )))
+    }
+}
+
+impl TryFrom<Message> for OpenPGPMessage {
+    type Error = Error;
+
+    fn try_from(value: Message) -> Result<Self, Self::Error> {
+        Self::try_from(&value)
+    }
+}
+
+impl TryFrom<OpenPGPMessage> for Message {
+    type Error = Error;
+
+    fn try_from(value: OpenPGPMessage) -> Result<Self, Self::Error> {
+        Ok(serde_json::from_str(
+            &value
+                .decompress()?
+                .get_literal()
+                .map_or_else(|| Err(Error::InvalidMessage), Ok)?
+                .to_string()
+                .map_or_else(|| Err(Error::InvalidMessage), Ok)?,
+        )?)
+    }
+}
+
+// Unsafe, here for reference if for some reason this is needed (appears to work with `pgp = "0.7.1"`)
+// This was only thought of because the signature field in [`StandaloneSignature`] was private in the latest version of rpgp (at the time "0.7.1").
+//
+// Update:
+// Found a safe workaround:
+// ```rust
+// if let pgp::composed::message::Message::Signed {
+//     message,
+//     one_pass_signature: _,
+//     signature,
+// } = message
+// { /* use signature here */ }
+// ```
+//
+// Old unsafe workaround:
+// ```rust
+// pub trait GetSignature {
+//     fn get_signature(self) -> Signature;
+// }
+
+// impl GetSignature for StandaloneSignature {
+//     fn get_signature(self) -> Signature {
+//         struct StandaloneSignature {
+//             signature: Signature,
+//         }
+//         let transmuted: StandaloneSignature = unsafe { std::mem::transmute(self) };
+//         transmuted.signature
+//     }
+// }
+// ```
+
+pub async fn get_or_import_public_key(
+    fingerprint: &str,
+    key_server: &str,
+) -> Result<SignedPublicKey> {
+    let file_name = format!(
+        "{}{}.asc",
+        USER_PUBLIC_KEY_FOLDER,
+        fingerprint.to_ascii_uppercase()
+    );
+    let path = std::path::Path::new(&file_name);
+    if path.is_file() {
+        Ok(SignedPublicKey::from_string(&tokio::fs::read_to_string(path).await?)?.0)
+    } else {
+        let url = format!(
+            "{}/pks/lookup?op=get&options=mr&search={}",
+            key_server, fingerprint
+        );
+        let response = reqwest::get(url).await?;
+
+        if response.status() == StatusCode::OK {
+            let text = response.text().await?;
+            let key = SignedPublicKey::from_string(&text)?.0;
+            let downloaded_fp = hex::encode_upper(key.fingerprint());
+            if downloaded_fp == fingerprint {
+                tokio::fs::write(path, text).await?;
```

Comment:
Is there any risk of denial of service here if people spam the server with new users, filling up the disk space and using up IO?

Reply:
I guess that there is, I suppose a way around it would be to either delete old keys or the to allow a key to be downloaded and used once (or cached in ram) I will have to figure that out

Reply:
This will probably be handled in a separate issue (#25).